feat: Add task-aware embedding support

StoreksFeed opened the pull request:
## Description

Modern embedding models (e.g., Gemini and FRIDA) support asymmetric embeddings through task-specific prefixes, which significantly improve retrieval accuracy by generating different embeddings for queries versus documents. This PR implements this capability in LightRAG.

## Related Issues

N/A

## Changes Made

### Configuration & Documentation
- Added `EMBEDDING_DOCUMENT_PREFIX` and `EMBEDDING_QUERY_PREFIX` environment variables to **`lightrag/api/config.py`**
- Updated **`docs/DockerDeployment.md`** and **`env.example`** with new configuration options
- Added example demonstrating prefix usage: **`examples/unofficial-sample/lightrag_embedding_prefixes.py`**

### Core Infrastructure
- Enhanced `EmbeddingFunc` wrapper in **`lightrag/utils.py`** with `supports_context` parameter
- Updated `wrap_embedding_func_with_attrs` decorator to support context-aware functions
- Modified **`lightrag/operate.py`** to pass `context`

### Vector Storage Backends
Updated all storage implementations to use context parameter:
- **`lightrag/kg/faiss_impl.py`**
- **`lightrag/kg/milvus_impl.py`**
- **`lightrag/kg/mongo_impl.py`**
- **`lightrag/kg/nano_vector_db_impl.py`**
- **`lightrag/kg/postgres_impl.py`**
- **`lightrag/kg/qdrant_impl.py`**

### LLM Provider Bindings
Updated embedding functions with context support:
- **`lightrag/llm/openai.py`** - Prefix support
- **`lightrag/llm/ollama.py`** - Prefix support
- **`lightrag/llm/gemini.py`** - Automatic task_type selection
- **`lightrag/llm/jina.py`** - Automatic task selection
- **`lightrag/llm/hf.py`** - Prefix support

### Binding Options
- Updated `GeminiEmbeddingOptions` to support automatic task_type selection

### API Server
- Integrated prefix configuration into **`lightrag/api/lightrag_server.py`**
- Updated **`lightrag/api/utils_api.py`** splash screen to display prefix settings

## Checklist

- [x] Changes tested locally
- [ ] Code reviewed
- [x] Documentation updated (if necessary)
- [ ] Unit tests added (if applicable)

## Additional Notes

### Backward Compatibility
- Fully backward compatible - task is not injected unless explicitly asked
- Existing deployments without prefix configuration should continue to work unchanged
- Optional feature activated only when `EMBEDDING_DOCUMENT_PREFIX` or `EMBEDDING_QUERY_PREFIX` environment variables are set

StoreksFeed committed:
Add context-aware embedding support with prefixes

StoreksFeed committed:
Fix default jina embedding task

StoreksFeed committed:
Fix formatting

StoreksFeed committed:
Fix typos in sample script

StoreksFeed committed:
Addressed review suggestions by Codex, updated naming, implemented clear opt-in

StoreksFeed committed:
Removed fallback to "document" if no context is provided

StoreksFeed committed:
Updated missed call sites with asymmetric embeddings

StoreksFeed committed:
Updated Jina embeddings default task

danielaskdd commented:
Thank you for your interest and contributions to LightRAG. Given the significant changes introduced in version 1.5, we will review this PR after official release of v1.5.

LarFii committed:
fix(utils): auto-detect supports_asymmetric in wrap_embedding_func_with_attrs

The new ``supports_asymmetric`` flag has to be set explicitly on every
provider in this PR (openai/azure_openai/ollama/hf/gemini/jina). Any
third-party or user-defined embedding function that adds a ``context``
parameter but forgets to also pass ``supports_asymmetric=True`` would
silently have its ``context`` value dropped by ``EmbeddingFunc.__call__``
(only ``logger.debug`` fires), defeating the whole task-aware feature.

Inspect the wrapped function's signature and default ``supports_asymmetric``
to True iff the function declares a ``context`` parameter. Explicit kwargs
still win, so callers who deliberately want the legacy stripping behavior
(e.g. when wrapping a context-accepting function but they don't want it
honored) can still pass ``supports_asymmetric=False`` and get the previous
result.

Made-with: Cursor

LarFii committed:
fix(jina): default task=None so context can drive task selection

With ``task='text-matching'`` as the default, library users calling
``jina_embed.func(texts, context='query')`` directly never got the new
behavior because the ``if task is None`` branch was skipped — only the
server-side dispatch (which manually sets ``task=None``) saw the
context-aware behavior. Switching the default to ``None`` makes the two
call paths consistent.

Backward compatibility: callers without ``context`` still fall through to
``task='text-matching'`` (the documented backward-compat default), so users
who previously relied on plain ``jina_embed(texts)`` semantics see no
change.

Made-with: Cursor

LarFii committed:
fix(example): initialize rag=None to avoid NameError in finally block

If ``initialize_rag()`` raises before the assignment to ``rag``, the
``finally: if rag:`` block hits ``UnboundLocalError`` and masks the real
exception. Initialize ``rag = None`` at the top of ``main`` so the cleanup
branch is safe regardless of where the failure occurred.

Made-with: Cursor

LarFii committed:
test(embedding): add unit tests for task-aware/asymmetric embedding

11 fully-mocked tests covering the new feature surface end-to-end:

  * ``wrap_embedding_func_with_attrs`` auto-detects ``supports_asymmetric``
    from the wrapped function's signature (3 cases: present, absent,
    explicit override).
  * ``EmbeddingFunc.__call__`` strips ``context`` for legacy embeds
    (no ``TypeError``) and forwards it for context-aware embeds.
  * ``jina_embed`` selects ``retrieval.query`` / ``retrieval.passage``
    from ``context`` when ``task`` is left at the default, and explicit
    ``task`` still overrides.
  * ``gemini_embed`` selects ``RETRIEVAL_QUERY`` / ``RETRIEVAL_DOCUMENT``
    from ``context`` when ``task_type`` is left at the default, and explicit
    ``task_type`` still overrides.

Includes a fixture that clears ``_get_gemini_client``'s lru_cache between
tests so SDK patches don't bleed across cases.

All tests run in ~13s under pytest-asyncio, no live API calls.

Made-with: Cursor

LarFii commented:
@StoreksFeed thanks for this PR — the asymmetric/task-aware embedding design is solid and the per-storage `context` plumbing is well-covered. I reviewed it end-to-end (real `bge-large-en-v1.5` via OpenRouter for path A, mocked SDK for jina/gemini path B, all 7 vector storage backends for context propagation, plus a full backward-compat sweep) and pushed a small set of fixes + tests directly to your branch since you opted in to maintainer edits. Happy to revert any of it if you'd rather take a different approach — see the safety net at the end.

I had to **force-push** because your branch was dirty against current main (a `postgres_impl.py` instrumentation conflict from main's perf-timing changes). The rebase is included in the new history. Backup ref below in case you want to undo.

## What I pushed (4 commits on top of a clean rebase onto main)

| Commit | Type | Why |
|---|---|---|
| `e5852b0` `fix(utils): auto-detect supports_asymmetric in wrap_embedding_func_with_attrs` | usability fix | Every provider in this PR (openai/azure/ollama/hf/gemini/jina) had to manually set `supports_asymmetric=True`. Any third-party / user-defined embed function that adds a `context` parameter but forgets the flag would have its `context` silently dropped by `EmbeddingFunc.__call__` (only a `logger.debug` fires). Six lines + `inspect.signature` fallback removes this footgun. Explicit `supports_asymmetric=False` still wins. |
| `fd3bc75` `fix(jina): default task=None so context can drive task selection` | call-path consistency | Library users calling `jina_embed.func(texts, context='query')` directly never got the new behavior because `task='text-matching'` (the previous default) skipped the `if task is None` branch. Only the server-side dispatch (which manually sets `task=None`) saw context-aware behavior. New default is `None`, so the two call paths agree; the existing "no context, no override" case still falls through to `text-matching` for backward compat. Empirically reproduced before the fix in `tests/test_asymmetric_embedding.py`. |
| `af84351` `fix(example): initialize rag=None to avoid NameError in finally block` | bug fix | If `initialize_rag()` raises, the `finally: if rag:` block hits `UnboundLocalError` and masks the real exception. One-line `rag = None` at the top. |
| `3a3e156` `test(embedding): add unit tests for task-aware/asymmetric embedding` | testing | 11 fully-mocked tests covering: `wrap_embedding_func_with_attrs` auto-detection (3 cases), `EmbeddingFunc.__call__` legacy stripping + forwarding (2), `jina_embed` task selection from context (3 incl. explicit override), `gemini_embed` `task_type` selection (3 incl. explicit override). Includes an `lru_cache` clear fixture for `_get_gemini_client` so SDK patches don't bleed across cases. ~13s under pytest-asyncio, no live API calls. Would have caught both fixes above at PR time. |

Net stat across the 4 added commits: `+281 / -10`.

## Verification

- Local pytest: `tests/test_asymmetric_embedding.py` 11/11 pass; pre-existing `tests/test_chunking.py` 43/43 still pass; full suite untouched.
- Real OpenRouter e2e re-run on the new HEAD with `bge-large-en-v1.5` and document/query prefixes set — 8/8 indexing inputs prefixed with `passage: `, 1/1 query input prefixed with `query: `, no-prefix session emits clean text. Confirmed the spy at `openai.resources.embeddings.AsyncEmbeddings.create`.
- Vector-storage `context` propagation re-verified end-to-end through `nano_vector_db_impl`: 13/13 indexing calls receive `context="document"`, 1/1 query call receives `context="query"`. Legacy embed (no `context` param, `supports_asymmetric=False`) sees 26 calls all clean — wrapper strips `context`, no `TypeError`.
- Auto-detection lands cleanly: `openai_embed.supports_asymmetric == True` (signature has `context`), and the new sample test confirms a hypothetical legacy embed without `context` correctly resolves to `False`.

## Conflict resolution during rebase

Three trivial conflicts during `git rebase main` (resolved manually, then `git rebase --continue`):

1. `lightrag/operate.py` — main refactored `_perform_kg_search` to batch multiple embedding kinds (`query`/`ll`/`hl`) into one call. Applied your `context="query"` to the new batched call site.
2. `lightrag/kg/postgres_impl.py` — main added `performance_timing_log(...)` instrumentation around the embedding step. Kept the timing AND your `context="document"`.
3. `lightrag/api/config.py` + `env.example` — both you and main appended in the same nearby region (file-upload-size args / new EMBEDDING examples). Trivial both-keep merge.

## Safety net — undoing anything

I created `backup/pr-2560-prereview-2026-04-25` on `HKUDS/LightRAG` pointing at your previous head `84f8f908`, so the original commit can be restored at any time:

```bash
# Restore from backup ref to your fork (also requires force):
gh api -X PATCH repos/StoreksFeed/LightRAG/git/refs/heads/embedding-prefixes \
  -f sha=84f8f908f483daa74ce4e54f9d321c748e4a7b51 -F force=true
```

You can also revert individual commits without touching the rest:

```bash
git revert e5852b0d  # auto-detect change
git revert fd3bc75c  # jina default task
git revert af843515  # example finally
git revert 3a3e1565  # tests
```

## Items I deliberately did NOT touch

- **Coupling with #2484 (Voyage)**: when both PRs land, the auto-merge is syntactically clean but the `voyageai` dispatch in `lightrag_server.py` doesn't receive the new `context`/prefix plumbing. Voyage's `input_type="query"|"document"` is the perfect target for your `supports_asymmetric` design — recommend a small follow-up PR after both merge that flips `voyageai_embed` to `supports_asymmetric=True` and maps `context` → `input_type` in the server dispatch. I commented similarly on #2484; happy to open the follow-up.
- **OpenAI / HF prefix participating in tiktoken truncation**: a real but minor effect; documented mention in docstring would be nice but not a blocker.
- **`utils_api` splash showing `EMBEDDING_ASYMMETRIC` flag**: cosmetic; left as a follow-up.

If anything here doesn't match your intent, push back / revert any commit and I'll adapt. Thanks again for the great PR!

LarFii committed:
style: apply ruff-format and remove unused import in tests

CI uses ruff v0.6.4 (pinned in .pre-commit-config.yaml). Reformat to match
that version's defaults and drop the unused ``AsyncMock`` import.

Made-with: Cursor
